table/block_based/block.cc

```diff
@@ -188,7 +188,7 @@ void DataBlockIter::Prev() {
     const Slice current_key(key_ptr, current_prev_entry.key_size);
 
     current_ = current_prev_entry.offset;
-    key_.SetKey(current_key, false /* copy */);
+    raw_key_.SetKey(current_key, false /* copy */);
```

Comment:
there's a bug here. the `current_key` came from the entries cached during Prev. Those cached keys have global seqno already assigned. If we assign such a key to `raw_key_` and then the scan direction switches to forward, then we're exposed to the delta-decoding bug we originally set out to avoid.

Reply:
Fixed. Changed Prev's caching to cache raw keys.

Reply:
I think this also fixes an existing bug when `ParseNextDataKey()` uses a cached `Prev()` entry as its starting point. In that case the seqno/value type it restores did not necessarily come from the current key. They came from the most recent key on which we called ParseNextDataKey(), which could be a later key. That means `ParseNextDataKey()` is restoring the wrong key's `ValueType` before it advances. Technically it's restoring the wrong key's seqno too, but it just so happens to work since in ingested files the original seqnos are all the same (zero).

Reply:
The existing bug is a bit different from what I described above. The restoration of the wrong key's seqno/value type just so happens to be OK because it restores them to the wrong buffer (oops). In that case, we're not restoring anything to the cached Prev entry before we use it for delta-decoding, which means the seqno was corrupted by the global seqno.